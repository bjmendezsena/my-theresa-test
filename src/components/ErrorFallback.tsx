export const ErrorFallback = () => {
  return <div>Ooops, something went wrong :(</div>;
};
