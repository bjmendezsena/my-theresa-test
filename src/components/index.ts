export * from './BackButton';
export * from './CategoryCarousel';
export * from './ErrorFallback';
export * from './ExampleComponent';
export * from './FavouriteMoviesGrid';
export * from './GenreList';
export * from './Hero';
export * from './Layouts';
export * from './MovieCard';
export * from './MovieInfo';
export * from './MovieMeta';
export * from './MovieOverview';
export * from './MoviePoster';
export * from './WishlistButton';
