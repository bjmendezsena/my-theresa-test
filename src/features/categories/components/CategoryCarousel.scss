.category-carousel {
  margin-bottom: $spacing-12;

  &__title {
    font-size: $text-2xl;
    font-weight: $font-bold;
    color: $white;
    margin: 0 0 $spacing-6;
    padding: 0 $spacing-6;

    @include mobile-only {
      font-size: $text-xl;
      padding: 0 $spacing-4;
      margin-bottom: $spacing-4;
    }
  }

  &__container {
    position: relative;
  }

  &__items {
    display: flex;
    gap: $spacing-3;
    overflow-x: auto;
    scroll-behavior: smooth;
    padding: 0 $spacing-6;
    scrollbar-width: none;
    -ms-overflow-style: none;

    &::-webkit-scrollbar {
      display: none;
    }

    @include mobile-only {
      padding: 0 $spacing-4;
      gap: $spacing-2;
    }

    @include tablet-up {
      padding: 0 $spacing-8;
      gap: $spacing-3;
    }

    @include desktop-up {
      padding: 0 $spacing-12;
      gap: $spacing-4;
    }
  }

  &__item {
    flex-shrink: 0;
    max-width: 250px;
    min-width: 180px;

    @include mobile-only {
      min-width: 140px;
      max-width: 160px;
    }

    @include tablet-up {
      min-width: 200px;
      max-width: 240px;
    }

    @include desktop-up {
      min-width: 220px;
      max-width: 280px;
    }

    @include large-desktop-up {
      max-width: 320px;
    }
  }

  &__empty {
    color: $gray-400;
    font-size: $text-base;
    text-align: center;
    padding: $spacing-8 $spacing-4;
    font-style: italic;
  }

  &__nav-btn {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    width: 50px;
    height: 50px;
    border: none;
    background: linear-gradient(
      135deg,
      rgba($black, 0.8),
      rgba($gray-900, 0.9)
    );
    color: $white;
    border-radius: $radius-full;
    cursor: pointer;
    transition: all $duration-300 $ease-out;
    backdrop-filter: blur(10px);
    box-shadow: $shadow-lg;
    z-index: $z-dropdown;
    display: flex;
    align-items: center;
    justify-content: center;

    &:hover:not(&--disabled) {
      background: linear-gradient(
        135deg,
        rgba($primary-600, 0.9),
        rgba($primary-700, 0.9)
      );
      transform: translateY(-50%) scale(1.1);
      box-shadow: $shadow-xl;
    }

    &:active:not(&--disabled) {
      transform: translateY(-50%) scale(0.95);
    }

    &--left {
      left: $spacing-2;

      @include mobile-only {
        left: $spacing-1;
      }
    }

    &--right {
      right: $spacing-2;

      @include mobile-only {
        right: $spacing-1;
      }
    }

    &--disabled {
      opacity: 0;
      pointer-events: none;
      transform: translateY(-50%) scale(0.8);
    }

    @include mobile-only {
      width: 40px;
      height: 40px;
    }
  }

  &__nav-icon {
    font-size: $text-2xl;
    font-weight: $font-bold;
    line-height: 1;
    user-select: none;

    @include mobile-only {
      font-size: $text-xl;
    }
  }

  &__container:hover {
    .category-carousel__nav-btn:not(.category-carousel__nav-btn--disabled) {
      opacity: 1;
    }
  }

  @include tablet-up {
    &__title {
      padding: 0 $spacing-8;
    }

    &__nav-btn {
      &--left {
        left: $spacing-3;
      }

      &--right {
        right: $spacing-3;
      }
    }
  }

  @include desktop-up {
    &__title {
      padding: 0 $spacing-12;
    }

    &__nav-btn {
      &--left {
        left: $spacing-4;
      }

      &--right {
        right: $spacing-4;
      }
    }
  }
}

@media (prefers-reduced-motion: no-preference) {
  .category-carousel__items {
    scroll-behavior: smooth;
  }
}

.category-carousel__nav-btn:focus {
  outline: 2px solid $primary-400;
  outline-offset: 2px;
}

.category-carousel--loading {
  .category-carousel__items {
    display: flex;
    gap: $spacing-3;
  }

  .category-carousel__item {
    width: calc((100vw - 2 * $spacing-6 - 4 * $spacing-4) / 3);
    height: 350px;
    background: linear-gradient(
      90deg,
      $gray-800 25%,
      $gray-700 50%,
      $gray-800 75%
    );
    background-size: 200% 100%;
    animation: shimmer 1.5s infinite;
    border-radius: $radius-lg;
    flex-shrink: 0;

    @include mobile-only {
      width: calc((100vw - 2 * $spacing-4 - 3 * $spacing-3) / 2.5);
      height: 280px;
    }

    @include tablet-up {
      width: calc((100vw - 2 * $spacing-8 - 4 * $spacing-4) / 3);
      height: 320px;
    }

    @include desktop-up {
      width: calc((100vw - 2 * $spacing-12 - 4 * $spacing-4) / 3);
      height: 380px;
    }

    @include large-desktop-up {
      width: calc((1200px - 4 * $spacing-4) / 3);
      height: 400px;
    }
  }
}

@keyframes shimmer {
  0% {
    background-position: 200% 0;
  }
  100% {
    background-position: -200% 0;
  }
}
