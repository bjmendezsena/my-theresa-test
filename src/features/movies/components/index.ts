export * from './BackButton';
export * from './GenreList';
export * from './Hero';
export * from './MovieInfo';
export * from './MovieMeta';
export * from './MovieOverview';
export * from './MoviePoster';
export * from './WishlistButton';
export * from './MovieCard';
