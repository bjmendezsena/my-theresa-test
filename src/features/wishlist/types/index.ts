export type WishlistItem = {
  id: number;
  title: string;
  posterPath: string;
  releaseDate: Date;
  overview: string;
  rating: number;
};
