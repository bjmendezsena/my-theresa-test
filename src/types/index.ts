import { QueryClient } from '@tanstack/react-query';

export type PrefetchFunction = (
  queryClient: QueryClient,
  params?: Record<string, any>
) => Promise<{
  queryClient: QueryClient;
  queryData?: any;
}>;
