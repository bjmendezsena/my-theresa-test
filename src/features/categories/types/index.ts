import { MovieItem } from '@/features/movies';

export type Category = {
  id: number;
  name: string;
  movies: MovieItem[];
};
