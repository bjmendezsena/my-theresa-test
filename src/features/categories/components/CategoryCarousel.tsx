import { useState, useRef, useEffect } from 'react';
import { Category } from '@/features/categories';
import { useDiscoverMovies } from '@/features/movies';
import { IMAGE_BASE_URL, MovieCard } from '@/features/movies';
import './CategoryCarousel.scss';

export interface CategoryCarouselProps {
  category: Category;
}

export const CategoryCarousel = ({ category }: CategoryCarouselProps) => {
  const { data } = useDiscoverMovies({
    genres: [category.id],
  });

  const { results = [] } = data || {};
  const carouselRef = useRef<HTMLDivElement>(null);
  const [canScrollLeft, setCanScrollLeft] = useState(false);
  const [canScrollRight, setCanScrollRight] = useState(true);

  const scrollAmount = 300; // Amount to scroll in pixels

  const updateScrollButtons = () => {
    if (!carouselRef.current) return;

    const { scrollLeft, scrollWidth, clientWidth } = carouselRef.current;
    setCanScrollLeft(scrollLeft > 0);
    setCanScrollRight(scrollLeft < scrollWidth - clientWidth - 1);
  };

  const scrollLeft = () => {
    if (!carouselRef.current) return;
    carouselRef.current.scrollBy({
      left: -scrollAmount,
      behavior: 'smooth',
    });
  };

  const scrollRight = () => {
    if (!carouselRef.current) return;
    carouselRef.current.scrollBy({
      left: scrollAmount,
      behavior: 'smooth',
    });
  };

  useEffect(() => {
    const carousel = carouselRef.current;
    if (!carousel) return;

    // Initial check
    updateScrollButtons();

    // Listen for scroll events
    carousel.addEventListener('scroll', updateScrollButtons);

    // Listen for resize events
    window.addEventListener('resize', updateScrollButtons);

    return () => {
      carousel.removeEventListener('scroll', updateScrollButtons);
      window.removeEventListener('resize', updateScrollButtons);
    };
  }, [results]);

  if (results.length === 0) {
    return (
      <div className="category-carousel">
        <h2 className="category-carousel__title">{category.name}</h2>
        <p className="category-carousel__empty">
          No movies found in this category.
        </p>
      </div>
    );
  }

  return (
    <div className="category-carousel">
      <h2 className="category-carousel__title">{category.name}</h2>

      <div className="category-carousel__container">
        <button
          className={`category-carousel__nav-btn category-carousel__nav-btn--left ${
            !canScrollLeft ? 'category-carousel__nav-btn--disabled' : ''
          }`}
          onClick={scrollLeft}
          disabled={!canScrollLeft}
          aria-label="Scroll left"
        >
          <span className="category-carousel__nav-icon">‹</span>
        </button>

        <div
          ref={carouselRef}
          className="category-carousel__items"
          onScroll={updateScrollButtons}
        >
          {results.map(movie => (
            <MovieCard
              key={movie.id}
              id={movie.id}
              title={movie.title}
              posterPath={`${IMAGE_BASE_URL}${movie.poster_path}`}
              releaseDate={movie.release_date?.toString()}
              voteAverage={movie.vote_average}
              overview={movie.overview}
              className="category-carousel__item"
            />
          ))}
        </div>

        <button
          className={`category-carousel__nav-btn category-carousel__nav-btn--right ${
            !canScrollRight ? 'category-carousel__nav-btn--disabled' : ''
          }`}
          onClick={scrollRight}
          disabled={!canScrollRight}
          aria-label="Scroll right"
        >
          <span className="category-carousel__nav-icon">›</span>
        </button>
      </div>
    </div>
  );
};
