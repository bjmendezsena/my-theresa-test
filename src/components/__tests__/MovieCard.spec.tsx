import _merge from 'lodash/merge';
import {
  renderWithProviders,
  expect,
  describe,
  test,
  act,
} from '@/tests/utils';
import { MovieCard, MovieCardProps } from '@/components';
import { wishlistStore } from '@/store';
import { WishlistItem } from '@/types';

const item: WishlistItem = {
  id: 1,
  overview: 'overview',
  posterPath: 'path/to/poster.jpg',
  rating: 5,
  releaseDate: '2023-01-01',
  title: 'Test Movie',
};
const defaultProps: MovieCardProps = {
  ...item,
  voteAverage: item.rating,
  showWishlistButton: true,
};

describe(`<${MovieCard.name}/>`, () => {
  const getComponent = (props: Partial<MovieCardProps> = {}) =>
    renderWithProviders(<MovieCard {..._merge(defaultProps, props)} />);

  test('renders correctly', () => {
    const { container } = getComponent();
    expect(container).toMatchSnapshot();
  });

  test('adds item to wishlist on click', () => {
    const { getByRole } = getComponent();
    const addButton = getByRole('button', { name: /add to wishlist/i });

    act(() => {
      addButton.click();
    });

    const wishlistItems = wishlistStore.getState().items;
    expect(wishlistItems).toHaveLength(1);
    expect(wishlistItems[0]).toEqual(item);
  });
});
