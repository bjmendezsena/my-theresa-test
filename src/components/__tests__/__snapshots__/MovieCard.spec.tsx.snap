// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<MovieCard/> > renders correctly 1`] = `
<div>
  <a
    aria-label="View details for Test Movie"
    class="movie-card "
    data-discover="true"
    href="/movie/1"
  >
    <div
      class="movie-card__poster"
    >
      <img
        alt="Test Movie poster"
        class="movie-card__poster-image"
        loading="lazy"
        src="https://image.tmdb.org/t/p/w500/path/to/poster.jpg"
      />
      <div
        class="movie-card__rating"
      >
        <span
          class="movie-card__rating-icon"
        >
          ‚≠ê
        </span>
        <span
          class="movie-card__rating-value"
        >
          5.0
        </span>
      </div>
      <button
        aria-label="Add to wishlist"
        class="movie-card__wishlist-btn "
        title="Add to wishlist"
      >
        <span
          class="movie-card__wishlist-icon"
        >
          ü§ç
        </span>
      </button>
    </div>
    <div
      class="movie-card__content"
    >
      <h3
        class="movie-card__title"
        title="Test Movie"
      >
        Test Movie
      </h3>
      <p
        class="movie-card__year"
      >
        2023
      </p>
      <p
        class="movie-card__overview"
      >
        overview
      </p>
    </div>
  </a>
</div>
`;
